dali/imgcodec/util/convert.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <utility>
+#include "dali/core/static_switch.h"
+#include "dali/imgcodec/util/convert.h"
+#include "dali/kernels/slice/slice_cpu.h"
+#include "dali/kernels/common/utils.h"
+
+namespace dali {
+namespace imgcodec {
+
+template <int static_ndim = -1, typename Out, typename In, typename ConvertFunc>
+void Convert(Out *out, const int64_t *out_strides,
+             const In *in, const int64_t *in_strides,
+             const int64_t *size, int ndim,
+             ConvertFunc &&func) {
+  if constexpr (static_ndim < 0) {
+    VALUE_SWITCH(ndim, NDim, (0, 1, 2, 3, 4),
+      (Convert<NDim>(out, out_strides, in, in_strides, size, NDim,
+                     std::forward<ConvertFunc>(func));
+      return;), ()
+    );  // NOLINT
+  }
+
+  int64_t extent = size[0];
+  int64_t in_stride = in_strides[0];
+  int64_t out_stride = out_strides[0];
+
+  if constexpr (static_ndim == 0) {
+    func(out, in);
+  } else if constexpr (static_ndim == 1) {  // NOLINT - if constexpr not recognized
+    for (int64_t i = 0; i < extent; i++) {
+      func(out + i * out_stride, in + i * in_stride);
+    }
+  } else {
+    assert(ndim != 1 && "This should go with the static ndim codepath");
+    for (int64_t i = 0; i < extent; i++) {
+      const int next_ndim = static_ndim < 0 ? -1 : static_ndim - 1;
+      Convert<next_ndim>(out + i * out_stride, out_strides + 1,
+                         in + i * in_stride, in_strides + 1,
+                         size + 1, ndim - 1,
+                         std::forward<ConvertFunc>(func));
+    }
+  }
+}
+
+template <typename Out, typename In>
+void ConvertDType(Out *out, const In *in) {
+  *out = ConvertSatNorm<Out>(*in);
+}
+
+#define IMG_CONVERT_TYPES uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, float, float16
+
+void Convert(SampleView<CPUBackend> out, TensorLayout out_layout, DALIImageType out_format,
+             ConstSampleView<CPUBackend> in, TensorLayout in_layout, DALIImageType in_format,
+             TensorShape<> roi_start, TensorShape<> roi_end) {
+  TensorShape<> out_shape = out.shape();
+  TensorShape<> in_shape = in.shape();
+  assert(in_shape.sample_dim() == out_shape.sample_dim());
+  assert(in_shape.sample_dim() == in_shape.sample_dim());
+  int ndim = out_shape.sample_dim();
+  int spatial_ndim = ndim - 1;
```

Comment:
Check that we actually have a channel dim?

Reply:
Look one line further.

Reply:
True